# JHU EN.600.749: Computational Genomics: Applied Comparative Genomics
## Project Presentations

Presentations will be a total of 15 minutes: 12 minutes for the presentation, followed by 3 minutes for questions. We will strictly keep to the schedule to ensure that all groups can present in class. 

## Schedule of Presentations

Day     | Time      | Team Name               | Students                                   | Title 
--------|-----------|-------------------------|--------------------------------------------|---------------------------------------------------------------------
Th 4/26 | 1:30-1:45 | Vishwanath              | Pranay Vishwanath                          | ChromHMM on plant genomes
Th 4/26 | 1:45-2:00 | miniMSA                 | Nathan Roach                               | Minimizer accelerated clustering and MSA of highly similar sequences 
Th 4/26 | 2:00-2:15 | Chromo                  | Ashlee Feng                                | Identifying Sequence-Specific Open Chromatin Regulators and Applications on Noncoding Variant Analysis
Th 4/26 | 2:15-2:30 | ZX Team                 | Zhuwei Xu                                  | Analysis of the Performance of MinMash algorithm for Copy Number Analysis by Pacbio reads in C. glabrata
Th 4/26 | 2:30-2:45 | The inTIMPidators       | Roham Razaghi, Timothy Gilpatrick          | Using nanopore sequencing data to examine Allele-specific expression and methylation patterns 
   .    |     .     |            .            |                       .                    |                              .
Tu 5/2  | 1:30-1:45 | Chesapeake Bay Explorer | Zhuoyue Zhang, Junyao Gu                   | A metagenomic-based survey of microbial (de)halogenation potential in the Chesapeake Bay
Tu 5/2  | 1:45-2:00 | Binomics                | Benjamin Kaminow                           | Analysis of Repeat Bits in Binaries Using DNA Short Read Assembly 
Tu 5/2  | 2:00-2:15 | Cao                     | Raymon Cao                                 | Evaluating Metagenomic Classification Software for Identifying Food Sample Contaminations
Tu 5/2  | 2:15-2:30 | Padhy                   | Shreyas Padhy                              | Deep Recurrent Networks for Base-Calling 
Tu 5/2  | 2:30-2:45 | Ray Charles Driving School Class of 2020 |Diego Gelsinger, Michael Skaro, Andrew Gale |Benchmarking of RNA-seq Analysis Pipelines using Circulating Tumor Expression Data
   .    |     .     |            .            |                       .                    |                              .
Th 5/4  | 1:30-1:45 | Araba-cadabra           | Michael Alonge                             | Pairwise Pan-Genome Structural Variant Detection in _Arabidopsis thaliana_ Genomes 
Th 5/4  | 1:45-2:00 | Xi                      | Wang Xi                                    | Systemic evaluation of ChromHMM and Segway, and chromatin state analysis on diploid human genome
Th 5/4  | 2:00-2:15 | ReAlign                 | Ales Varabyou                              | Graph genome assisted realignment of short-read sequencing data for small highly divergent genomes. 
Th 5/4  | 2:15-2:30 | KrisprKreme             | Srivathsa Pasumarthi                       | On-target activity prediction of sgRNA in CRISPR/CAS9 using Deep Neural Networks
Th 5/4  | 2:30-2:45 | Insertion Short         | Michael Kirsche                            | Short-read Known Insertion Detection 
Th 5/4  | 2:45-3:00 | Gray                    | Justin Gray                                | Finding Gene Regulatory Pathways Shared Between Dormant Tumor Cells 
    
<br>
<br>

## Recommended outline for your talk (1 minute per slide):

1. Title Slide: Who are you, title, date
2. Intro 1: Whats the big idea???
3. Intro 2: More specifically, what are you trying to learn?
4. Methods 1: What did you try?
5. Methods 2: What is the key idea?
6. Data 1: What data are you looking at?
7. Data 2: Anything notable about the data?
8. Results 1: What did you see!
9. Results 2: Does it work?
10. Results 3: How does it compare to other methods/data/ideas?
11. Discussion 1: What did you learn from this study?
12. Discussion 2: What does this mean for the future?
13. Acknowledgements: Who helped you along the way?

I strongly *discourage* you from trying to give a live demo as they are too unpredictable for a short talk. If you have running software you want to show, use a "cooking show" approach, where you have screen shots of the important steps.    
